components/vision.py

```diff
@@ -1,213 +1,161 @@
 import math
-from components.chassis import Chassis
-import wpilib
-import wpiutil.log
-import robotpy_apriltag
-from wpimath.geometry import Transform3d, Translation3d, Rotation3d, Pose2d, Quaternion
+import typing
 from typing import Optional
-from magicbot import tunable, feedback
-
-from photonvision import (
-    PhotonCamera,
-    PhotonTrackedTarget,
-)
-
-
-class Vision:
-    chassis: Chassis
-
-    FIELD_LAYOUT = robotpy_apriltag.loadAprilTagLayoutField(
-        robotpy_apriltag.AprilTagField.k2023ChargedUp
-    )
 
-    # TBD
-    X_STD_DEV_CONSTANT = Y_STD_DEV_CONSTANT = 3.0
-    ANGULAR_STD_DEV_CONSTANT = 1.0
+import wpilib
+import wpiutil.log
+from magicbot import tunable
+from photonvision import PhotonCamera, PhotonTrackedTarget
+from wpimath.geometry import Pose2d, Rotation3d, Transform3d, Translation3d
 
-    field: wpilib.Field2d
+from components.chassis import Chassis
+from utilities.game import apriltag_layout
+
+
+class VisualLocalizer:
+    """
+    This localizes the robot from AprilTags on the field,
+    using information from a single PhotonVision camera.
+    """
+
+    add_to_estimator = tunable(False)
+    should_log = tunable(False)
+
+    rejected_in_row = tunable(0)
+    last_pose_z = tunable(0.0, writeDefault=False)
+
+    def __init__(
+        self,
+        # The name of the camera in PhotonVision.
+        name: str,
```

Comment:
New feature in magicbot which I still haven't written docs for: it does dependency injection into the constructor :D